test/libyul/ControlFlowSideEffectsTest.cpp

```diff
@@ -61,19 +62,20 @@ TestCase::TestResult ControlFlowSideEffectsTest::run(ostream& _stream, string co
 	if (!obj.code)
 		BOOST_THROW_EXCEPTION(runtime_error("Parsing input failed."));
 
-	std::map<YulString, ControlFlowSideEffects> sideEffects =
-		ControlFlowSideEffectsCollector(
-			EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
-			*obj.code
-		).functionSideEffects();
-
-	std::map<std::string, std::string> controlFlowSideEffectsStr;
-	for (auto&& [fun, effects]: sideEffects)
-		controlFlowSideEffectsStr[fun.str()] = toString(effects);
-
-	m_obtainedResult.clear();
-	for (auto&& [functionName, effect]: controlFlowSideEffectsStr)
-		m_obtainedResult += functionName + (effect.empty() ? ":" : ": " + effect) + "\n";
+	ControlFlowSideEffectsCollector sideEffects(
+		EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
+		*obj.code
+	);
+	std::vector<std::string> result;
+	for (auto&& [function, effects]: sideEffects.functionSideEffects())
+	{
+		string effectStr = toString(effects);
+		result.emplace_back(
+			function->name.str() + (effectStr.empty() ? ":" : ": " + effectStr) + "\n"
+		);
+	}
+	sort(result.begin(), result.end());
+	m_obtainedResult = util::joinHumanReadable(result, "");
```

Comment:
```suggestion
struct FunctionCollector: ASTWalker
{
	void operator()(FunctionDefinition const& _fun) override
	{
		funs.emplace_back(_fun);
		ASTWalker::operator()(_fun);
	}
	vector<FunctionDefinition const*> functions;
};
FunctionCollector functionCollector;
functionCollector(*obj.code);
for (FunctionDefinition const* fun: functionCollector.functions)
{
	string effectStr = toString(sideEffects.functionSideEffects().at(fun));
	  m_obtainedResult += fun->name.str() + (effectStr.empty() ? ":" : ": " + effectStr) + "\n";
}
```

Can't we do it like this :-)? That's not *that* much more code and this way the expectations are actually meaningful in that I can tell which ``b`` is which and don't sort them by side effects...

Reply:
Or even
```
struct: ASTWalker
{
	ControlFlowSideEffectsCollector sideEffects;
	string& obtainedResult;
	void operator()(FunctionDefinition const& _fun) override
	{
		string effectStr = toString(sideEffects.functionSideEffects().at(fun));
		m_obtainedResult += fun->name.str() + (effectStr.empty() ? ":" : ": " + effectStr) + "\n";
		ASTWalker::operator()(_fun);
	}
	vector<FunctionDefinition const*> functions;
} createResult {
	ControlFlowSideEffectsCollector{
		EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
		*obj.code
	},
	obtainedResult;
}(*obj.code);
```

Reply:
We could even generalize this with
```C++
template <typename Node, typename Visitor>
struct ForEach: std::conditional_t<std::is_const_v<Node>, ASTWalker, ASTModifier> {
	Visitor visitor;
	template<typename Callable>
	ForEach(Callable&& _visitor): visitor(std::forward<Visitor>(_visitor)) {}
	void operator()(Node& _node) override
	{
		visitor(_node);
		std::conditional_t<std::is_const_v<Node>, ASTWalker, ASTModifier>::operator()(_node);
	}
	using std::conditional_t<std::is_const_v<Node>, ASTWalker, ASTModifier>::operator();
};
template<typename Node, typename Entry, typename Visitor>
void forEach(Entry&& _entry, Visitor&& visitor)
{
	ForEach<Node, Visitor>(std::forward<Visitor>(visitor))(std::forward<Entry>(_entry));
}
```

and then just use that:
```C++
	forEach<FunctionDefinition const>(*obj.code, [&](auto const& _fun) {
		string effectStr = toString(sideEffects.functionSideEffects().at(&_fun));
		m_obtainedResult += _fun.name.str() + (effectStr.empty() ? ":" : ": " + effectStr) + "\n";
	});
```
might help elsewhere...

Reply:
True!